### PR TITLE
Performance Enhancements

### PR DESCRIPTION
**Pull request by cb-anthony-merrill**
Contains the initial code for performance enhancements

### CODE REVIEW

File: kinsumer/checkpoints.go [hunk 1]
```diff
@@ -25,6 +25,7 @@ type checkpointer struct {
 	ownerName             string
 	ownerID               string
 	maxAgeForClientRecord time.Duration
+	stats                 StatReceiver
```
Comment (cb-xav) [inline kinsumer/checkpoints.go]: This is supposed to be removed.

File: kinsumer/checkpoints.go [hunk 2]
```diff
@@ -54,7 +55,8 @@ func capture(
 	dynamodbiface dynamodbiface.DynamoDBAPI,
 	ownerName string,
 	ownerID string,
-	maxAgeForClientRecord time.Duration) (*checkpointer, error) {
+	maxAgeForClientRecord time.Duration,
+	stats StatReceiver) (*checkpointer, error) {
```
Comment (cb-xav) [inline kinsumer/checkpoints.go]: Supposed to be removed.

File: kinsumer/checkpoints.go [hunk 3]
```diff
@@ -149,6 +151,7 @@ func capture(
 		dynamodb:              dynamodbiface,
 		ownerName:             ownerName,
 		ownerID:               ownerID,
+		stats:                 stats,
```
Comment (cb-xav) [inline kinsumer/checkpoints.go]: This is supposed to be removed.

File: kinsumer/checkpoints.go [hunk 4]
```diff
@@ -234,8 +239,11 @@ func (cp *checkpointer) commit() (bool, error) {
 	logger.WithFields(logrus.Fields{
 	    "sequenceNumber": sn,
 		"shardId": cp.shardID,
-	}).Info("Checkpoint committed")
+	}).Debug("Checkpoint committed")
 
+	if sn != nil {
```
Comment (cb-xav) [inline kinsumer/checkpoints.go]: Supposed to be removed.

File: kinsumer/checkpoints.go [hunk 5]
```diff
@@ -277,6 +285,10 @@ func (cp *checkpointer) release() error {
 		return fmt.Errorf("error releasing checkpoint: %s", err)
 	}
 
+	if cp.sequenceNumber != "" {
```
Comment (cb-xav) [inline kinsumer/checkpoints.go]: Supposed to be removed.

File: kinsumer/checkpoints_test.go [hunk 6]
```diff
@@ -12,13 +12,9 @@ import (
 func TestCheckpointer(t *testing.T) {
 	table := "checkpoints"
 	mock := mocks.NewMockDynamo([]string{table})
-	testChan := make(chan int)
-	seq1 := "seq1"
-	seq2 := "seq2"
-	seq3 := "seq3"
-	lastseq := "lastseq"
+	stats := &NoopStatReceiver{}
```
Comment (cb-xav) [inline kinsumer/checkpoints_test.go]: This whole file should not have updates.

File: kinsumer/config.go [hunk 7]
```diff
@@ -11,6 +11,8 @@ import (
 
 // Config holds all configuration values for a single Kinsumer instance
 type Config struct {
+	stats StatReceiver
```
Comment (cb-xav) [inline kinsumer/config.go]: This whole file should not have updates.

File: kinsumer/config_test.go [hunk 8]
```diff
@@ -47,15 +47,21 @@ func TestConfigErrors(t *testing.T) {
 	config = NewConfig().WithBufferSize(0)
 	err = validateConfig(&config)
 	require.EqualError(t, err, ErrConfigInvalidBufferSize.Error())
+
```
Comment (cb-xav) [inline kinsumer/config_test.go:50]: This whole file should not have updates.

File: kinsumer/errors.go [hunk 9]
```diff
@@ -30,6 +30,8 @@ var (
 	ErrConfigInvalidLeaderActionFrequency = errors.New("leaderActionFrequency config value is mandatory and must be at least as long as ShardCheckFrequency")
 	// ErrConfigInvalidBufferSize - BufferSize config value is mandatory
 	ErrConfigInvalidBufferSize = errors.New("bufferSize config value is mandatory")
+	// ErrConfigInvalidStats - Stats cannot be nil
+	ErrConfigInvalidStats = errors.New("stats cannot be nil")
```
Comment (cb-xav) [inline kinsumer/errors.go]: This file should not have updates.

File: kinsumer/kinsumer_test.go [hunk 10]
```diff
@@ -7,7 +7,7 @@ import (
 	"math/rand"
 	"sort"
 	"strconv"
-	//"sync"
+	"sync"
```
Comment (cb-xav) [inline kinsumer/kinsumer_test.go]: This whole file shouldn't have updates.

File: kinsumer/noopstatreceiver.go [hunk 11]
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) 2016 Twitch Interactive
```
Comment (cb-xav) [inline kinsumer/noopstatreceiver.go]: This file should not exist.

File: kinsumer/statreceiver.go [hunk 12]
```diff
@@ -0,0 +1,34 @@
+// Copyright (c) 2016 Twitch Interactive
```
Comment (cb-xav) [inline kinsumer/statreceiver.go]: This file should not exist.

File: kinsumer/statsd/statsd.go [hunk 13]
```diff
@@ -0,0 +1,67 @@
+// Copyright (c) 2016 Twitch Interactive
```
Comment (cb-xav) [inline kinsumer/statsd/statsd.go]: This file should not exist.

File: plugins/dest/emitKinesis.go [hunk 14]
```diff
@@ -1,26 +1,3 @@
-/*
-# The MIT License (MIT)
```
Comment (cb-xav) [inline plugins/dest/emitKinesis.go]: Is the license supposed to be gone?

File: plugins/dest/s3batch.go [hunk 15]
```diff
@@ -1,34 +1,9 @@
-/*
-# The MIT License (MIT)
-#
-# Copyright (c) 2019  Carbon Black
```
Comment (cb-xav) [inline plugins/dest/s3batch.go]: Is the license supposed to be gone?

File: ruleset/rule/rule.go [hunk 16]
```diff
@@ -1,26 +1,3 @@
-/*
-# The MIT License (MIT)
```
Comment (cb-xav) [inline ruleset/rule/rule.go]: Is the license supposed to be gone?

File: ruleset/rulebase/rulebase.go [hunk 17]
```diff
@@ -1,40 +1,13 @@
-/*
-# The MIT License (MIT)
```
Comment (cb-xav) [inline ruleset/rulebase/rulebase.go]: Is the license supposed to be gone?

File: ruleset/ruleset.go [hunk 18]
```diff
@@ -1,26 +1,3 @@
-/*
-# The MIT License (MIT)
-#
-# Copyright (c) 2019  Carbon Black
```
Comment (cb-xav) [inline ruleset/ruleset.go]: Is the license supposed to be gone?

File: ruleset/ruleset.go [hunk 19]
```diff
@@ -1021,21 +985,20 @@ func FindFuncRecursive(rule string, builder *bldr.Rule, isProject string, realID
...
-					"isMulti":         isMulti,
-					"wasMulti": wasMulti,
-				}).Debug("Pushing back Step - 2")
 				if isOr {
 					builder.OrPredicate(step, len(multiargs) < isMulti && isMulti > 1)
 				} else {
 					builder.AddStep(step, len(multiargs) < isMulti && isMulti > 1)
 				}
-			}
+				//fmt.Printf("Pushing back Step (%v) = %v\n", (*function).Name(), step.Value)
```
Comment (cb-xav) [inline ruleset/ruleset.go]: These commented lines can be removed.

File: s3Batcher/s3bufferclient.go [hunk 20]
```diff
@@ -1,26 +1,3 @@
-/*
-# The MIT License (MIT)
-#
-# Copyright (c) 2019  Carbon Black
```
Comment (cb-xav) [inline s3Batcher/s3bufferclient.go]: Is the license supposed to be removed?

**Review by cb-xav — COMMENTED** [review]
Lots of unnecessary stuff in here and lots of licenses removed.

File: plugins/dest/dummyDest.go [hunk 21]
```diff
@@ -45,7 +45,7 @@ func (g destpluginInterface) DoCheckpoint() bool {
 }
 
 func (g destpluginInterface) Publish(args ...interface{}) (result bool, err error) {
-	fmt.Printf("The value(s) to be sent downstream -> \n`%v`\n", string(*args[1].(*[]byte)))
+	fmt.Printf("The value(s) to be sent downstream -> \n`%v`\n", string(args[1].([]byte)))
```
Comment (cb-xav) [inline plugins/dest/dummyDest.go]: This should be a logger statement

**Review by cb-xav — COMMENTED** [review]
Logging

File: kinsumer/shard_consumer.go [hunk 22]
```diff
@@ -162,7 +165,8 @@ func (k *Kinsumer) captureShard(shardID string) (*checkpointer, error) {
 			k.dynamodb,
 			k.clientName,
 			k.clientID,
-			k.maxAgeForClientRecord)
+			k.maxAgeForClientRecord,
+			k.config.stats)
```
Comment (cb-xav) [inline kinsumer/shard_consumer.go]: Stats has to go.

File: kinsumer/shard_consumer.go [hunk 23]
```diff
@@ -539,9 +587,14 @@ mainloop:
 					}
 
 					// Run rules should take the inited rules above and execute those
-					if len(*builtRules) > 0 && len(payload) > 0 {
-						ruleset.RunRules(recordStruct, builtRules)
-						checkpointer.update(record.SequenceNumber, checkChan, chkRefs)
+					if len(*builtRules) > 0 && len(buffer) > 0  && recordStruct != nil {
+						//commented out because of http and uint8 error
```
Comment (cb-xav) [inline kinsumer/shard_consumer.go:590]: Want to update with go processRecords?

File: plugins/dest/dummyDest.go [hunk 24]
```diff
@@ -22,7 +48,12 @@ func (g destpluginInterface) DoCheckpoint() bool {
 }
 
 func (g destpluginInterface) Publish(args ...interface{}) (result bool, err error) {
-	fmt.Printf("The value(s) to be sent downstream -> \n`%v`\n", string(*args[1].(*[]byte)))
+	fmt.Printf("The value(s) to be sent downstream -> \n`%v`\n", )
```
Comment (cb-xav) [inline plugins/dest/dummyDest.go]: Does printf still need to be here?

File: kinsumer/shard_consumer.go (hunk 23 above)
Comment (cb-anthony-merrill) [reply kinsumer/shard_consumer.go:590]: Leaving this until we do our Enhanced Fanout push :)

**Review by cb-xav — APPROVED** [review]

**Review by mxiao-cb — APPROVED** [review]

**Review by dpat-cb — APPROVED** [review]

**Pull request merged**